### PR TITLE
[SYCL] Change the behavior of sycl::maximum and sycl::minimum to be consistent with std::max and std::min

### PR DESCRIPTION
**Pull request by eunakim0103**
The behavior of sycl::maximum and sycl::minimum used to be different from std::max and std::min in the case of one of two operand was NaN since sycl::maximum was implemented with std::greater<> while std::max used operator <. 
https://github.com/KhronosGroup/SYCL-Docs/pull/285

### CODE REVIEW

File: sycl/include/sycl/functional.hpp [hunk 1]
```diff
@@ -37,34 +38,35 @@ template <> struct logical_or<void> : std::logical_or<void> {};
 
 template <typename T = void> struct minimum {
   T operator()(const T &lhs, const T &rhs) const {
-    return std::less<T>()(lhs, rhs) ? lhs : rhs;
+    return (lhs > rhs) ? rhs : lhs;
```
Comment (Pennycook) [inline sycl/include/sycl/functional.hpp:41]:
```suggestion
    return (rhs < lhs) ? rhs : lhs;
```

File: sycl/include/sycl/functional.hpp [hunk 2]
```diff
@@ -37,34 +38,35 @@ template <> struct logical_or<void> : std::logical_or<void> {};
...
   template <typename T, typename U>
-  auto operator()(T &&lhs, U &&rhs) const -> std::common_type_t<T &&, U &&> {
-    return std::less<>()(std::forward<const T>(lhs), std::forward<const U>(rhs))
-               ? std::forward<T>(lhs)
-               : std::forward<U>(rhs);
+  auto operator()(T &&lhs, U &&rhs) const ->
+      typename std::common_type<T &&, U &&>::type {
+    return (std::forward<const T>(lhs) > std::forward<const U>(rhs))
+               ? std::forward<T>(rhs)
+               : std::forward<U>(lhs);
```
Comment (Pennycook) [inline sycl/include/sycl/functional.hpp:52]:
```suggestion
    return (std::forward<const U>(rhs) < std::forward<const T>(lhs))
               ? std::forward<U>(rhs)
               : std::forward<T>(lhs);
```

**Review by Pennycook — CHANGES_REQUESTED** [review]
To match C++ behavior, both minimum and maximum need to use the same operator (`<`) for ordering.  I think my suggestion for how to implement `minimum` is correct, following the logic outlined in https://github.com/KhronosGroup/SYCL-Docs/pull/285#issuecomment-1262485753.

File: sycl/include/sycl/functional.hpp [hunk 3]
```diff
@@ -23,6 +23,7 @@ template <typename T = void> using bit_xor = std::bit_xor<T>;
 
 // std:logical_and/std::logical_or with a non-void type returns bool,
 // sycl requires returning T.
+// these definition should be consistent with std::max, std::min
```
Comment (dm-vodopyanov) [inline sycl/include/sycl/functional.hpp:26]: These definitions / This definition ?

File: sycl/test/basic_tests/sycl_maximum_minimum.cpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
+// RUN: %clangxx -fsycl -fsyntax-only %s
```
Comment (dm-vodopyanov) [inline sycl/test/basic_tests/sycl_maximum_minimum.cpp:1]: Currently this test checks nothing. It has asserts in it, but this compilation line does not produce a binary, and binary execution line is missing.

File: sycl/include/sycl/functional.hpp (hunk 3 above)
Comment (eunakim0103) [reply sycl/include/sycl/functional.hpp:26]: It meant "sycl::minimum and sycl::maximum" by these. I will fix the comment

File: sycl/test/basic_tests/sycl_maximum_minimum.cpp (hunk 4 above)
Comment (eunakim0103) [reply sycl/test/basic_tests/sycl_maximum_minimum.cpp:1]: Let me fix this too. Thanks!

File: sycl/include/sycl/functional.hpp (hunk 3 above)
Comment (dm-vodopyanov) [reply sycl/include/sycl/functional.hpp:26]: Why it's marked as resolved? There is no fix in this PR.

File: sycl/test/basic_tests/sycl_maximum_minimum.cpp (hunk 4 above)
Comment (dm-vodopyanov) [reply sycl/test/basic_tests/sycl_maximum_minimum.cpp:1]: Why it's marked as resolved? There is no fix in this PR.

File: sycl/include/sycl/functional.hpp [hunk 5]
```diff
@@ -23,6 +23,7 @@ template <typename T = void> using bit_xor = std::bit_xor<T>;
 
 // std:logical_and/std::logical_or with a non-void type returns bool,
 // sycl requires returning T.
+
```
Comment (dm-vodopyanov) [inline sycl/include/sycl/functional.hpp:26]: Please remove this extra line.

File: sycl/include/sycl/functional.hpp [hunk 6]
```diff
@@ -35,36 +36,40 @@ template <typename T = void> struct logical_or {
 
 template <> struct logical_or<void> : std::logical_or<void> {};
 
+// sycl::minimum, sycl::maximum definition should be consistent with std::max,
```
Comment (dm-vodopyanov) [inline sycl/include/sycl/functional.hpp:39]: Looks like typos?

sycl::minimum --- std::max
sycl::maximum --- std::min

File: sycl/test/basic_tests/sycl_maximum_minimum.cpp [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+          (std::isnan(sycl::maximum{}(f2, f1)))) &&
+         "sycl::maximum result is wrong");
+  assert(((std::min(f1, f2) == sycl::minimum{}(f1, f2)) &&
+          "sycl::minimum result is wrong"));
+  assert(((std::isnan((std::min(f2, f1)))) &&
+          (std::isnan(sycl::minimum{}(f2, f1)))) &&
+         "sycl::minimum result is wrong");
+
+  return 0;
+}
```
Comment (dm-vodopyanov) [inline sycl/test/basic_tests/sycl_maximum_minimum.cpp:27]: Please restore the empty line.

File: sycl/test/basic_tests/sycl_maximum_minimum.cpp [hunk 8]
```diff
@@ -0,0 +1,27 @@
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %t.out
+
+#include <CL/sycl.hpp>
```
Comment (dm-vodopyanov) [inline sycl/test/basic_tests/sycl_maximum_minimum.cpp:4]: Let's align with SYCL 2020: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:headers-and-namespaces

File: sycl/include/sycl/functional.hpp [hunk 9]
```diff
@@ -35,36 +35,39 @@ template <typename T = void> struct logical_or {
 
 template <> struct logical_or<void> : std::logical_or<void> {};
 
+// sycl::minimum and sycl::maximum definition should be consistent
+// with std::min and std::max respectively
```
Comment (dm-vodopyanov) [inline sycl/include/sycl/functional.hpp:39]:
```suggestion
// sycl::minimum implementation should be consistent with std::min
```

File: sycl/include/sycl/functional.hpp [hunk 10]
```diff
@@ -35,36 +35,39 @@ template <typename T = void> struct logical_or {
...
-               : std::forward<U>(rhs);
+  auto operator()(T &&lhs, U &&rhs) const ->
+      typename std::common_type<T &&, U &&>::type {
+    return (std::forward<const U>(rhs) < std::forward<const T>(lhs))
+               ? std::forward<U>(rhs)
+               : std::forward<T>(lhs);
   }
 };
 
 template <typename T = void> struct maximum {
```
Comment (dm-vodopyanov) [inline sycl/include/sycl/functional.hpp:57]:
```suggestion
// sycl::maximum implementation should be consistent with std::max
template <typename T = void> struct maximum {
```

**Review by dm-vodopyanov — APPROVED** [review]

**Comment by eunakim0103** [comment]
Three test fails are unrelated to my PR changes: [AWS]CUDA LLVM Test Suite, HIP AMDGPU LLVM Test Suite, SYCL E2E on Intel CPU/GEN9 GPU

FAIL: SYCL :: Basic/accessor/empty_zero_dim_accessor.cpp (97 of 1445)
******************** TEST 'SYCL :: Basic/accessor/empty_zero_dim_accessor.cpp' FAILED ********************

**Comment by dm-vodopyanov** [comment]
@eunakim0103 please update the branch, there are too many failures on Intel CPU/GEN9:

```
Failed Tests (30):
  SYCL :: Basic/accessor/accessor.cpp
  SYCL :: Basic/accessor/empty_zero_dim_accessor.cpp
  SYCL :: Basic/get_backend.cpp
  SYCL :: Basic/get_info_aspect.cpp
  SYCL :: Basic/host_write_back.cpp
  SYCL :: Basic/queue_priority.cpp
  SYCL :: DeviceArchitecture/device_architecture_on_host.cpp
  SYCL :: DeviceLib/imf_fp32_test.cpp
  SYCL :: InvokeSimd/Regression/ImplicitSubgroup/slm_load_store.cpp
  SYCL :: InvokeSimd/Regression/slm_load_store.cpp
  SYCL :: Plugin/queue_submit_mode.cpp
  SYCL :: Regression/device_num.cpp
  SYCL :: Regression/empty_accessor_use.cpp
  SYCL :: Regression/exclusive-scan-char-short.cpp
  SYCL :: SpecConstants/2020/non-packed-struct.cpp
  SYCL :: Tracing/usm/queue_copy_released_pointer.cpp
  SYCL :: Tracing/usm/queue_single_task_nullptr.cpp
  SYCL :: Tracing/usm/queue_single_task_released_pointer.cpp
  SYCL :: XPTI/buffer/accessors.cpp
  SYCL :: XPTI/buffer/host_array.cpp
  SYCL :: XPTI/buffer/in_cycle.cpp
  SYCL :: XPTI/buffer/multiple_buffers.cpp
  SYCL :: XPTI/buffer/multiple_queues.cpp
  SYCL :: XPTI/buffer/recursion.cpp
  SYCL :: XPTI/buffer/sub_buffer.cpp
  SYCL :: XPTI/buffer/use_host_ptr.cpp
  SYCL :: XPTI/image/accessors.cpp
  SYCL :: XPTI/kernel/basic.cpp
  SYCL :: XPTI/kernel/content.cpp
  SYCL :: XPTI/mem_alloc_events_linux.cpp
```

**Review by Pennycook — APPROVED** [review]

**Pull request merged**